File: Python/Matrix.py
import random


class Matrix:
    def __init__(self, rows, cols):
        if type(rows) == "list":
            self.cols = len(rows[0])
            self.rows = rows.length
            self.data = rows
        else:
            self.cols = cols
            self.rows = rows
            self.data = []
            for i in range(self.rows):
                self.data.append([])
                for j in range(self.cols):
                    self.data[i].append(0)

    def print(self):
        print(self.data)

    @staticmethod
    def from_array(arr):
        m = Matrix(len(arr), 1)
        for i in range(len(arr)):
            m.data[i][0] = arr[i]
        return m

    @staticmethod
    def subtract(a, b):
        m = Matrix(a.rows, a.cols)
        for i in range(a.rows):
            for j in range(a.cols):
                m.data[i][j] = a.data[i][j] - b.data[i][j]
        return m

    @staticmethod
    def multiply(a, b):
        # dot product
        if a.cols == b.rows:
            m = Matrix(a.rows, b.cols)
            for i in range(a.rows):
                for j in range(b.cols):
                    total = 0
                    for k in range(b.rows):
                        total += a.data[i][k] * b.data[k][j]
                    m.data[i][j] = total
            return m
        else:
            raise ValueError("Columns should match rows of the matrix.")

    @staticmethod
    def transpose(m):
        # use cols in place of rows
        result = Matrix(m.cols, m.rows)
        for i in range(m.rows):
            for j in range(m.cols):
                result.data[j][i] = m.data[i][j]
        return result

    def randomize(self):
        for i in range(self.rows):
            for j in range(self.cols):
                self.data[i][j] = random.random() * 2 - 1

    def multiply_element_wise(self, m):
        if isinstance(m, Matrix):
            for i in range(self.rows):
                for j in range(self.cols):
                    self.data[i][j] *= m.data[i][j]
        else:
            for i in range(self.rows):
                for j in range(self.cols):
                    self.data[i][j] *= m

    def add(self, m):
        if isinstance(m, Matrix):
            for i in range(self.rows):
                for j in range(self.cols):
                    self.data[i][j] += m.data[i][j]
        else:
            for i in range(self.rows):
                for j in range(self.cols):
                    self.data[i][j] += m

    def map(self, df):
        m = Matrix(self.rows, self.cols)
        for i in range(self.rows):
            for j in range(self.cols):
                m.data[i][j] = df(self.data[i][j])
        return m

    def apply(self, df):
        for i in range(self.rows):
            for j in range(self.cols):
                self.data[i][j] = df(self.data[i][j])

    def to_array(self):
        array = []
        for i in range(self.rows):
            for j in range(self.cols):
                array.append(self.data[i][j])
        return array

File: Javascript/Matrix.js
class Matrix{
    constructor(rows, cols){
        if(typeof rows === "object"){
            this.cols = rows[0].length;
            this.rows = rows.length;
            this.data = rows;
        } else {
            this.cols = cols || 1;
            this.rows = rows || 1;
            this.data = [];
            for(let i = 0; i < this.rows; i++){
                this.data[i] = [];
                for(let j = 0; j < this.cols; j++){
                    this.data[i][j] = 0;
                }
            }
        }
        
    }
    static fromArray(arr){
        let m = new Matrix(arr.length, 1);
        for(let i = 0, len = arr.length; i < len; i++){
            m.data[i][0] = arr[i];
        }
        return m;
    }
    toArray(){
        let array = [];
        for(let i = 0; i < this.rows; i++){
            for(let j = 0; j < this.cols; j++){
                array.push(this.data[i][j]);
            }
        }
        return array;
    }
    
    static subtract(a, b){
        let result = new Matrix(a.rows, a.cols);
        for(let i = 0; i < a.rows; i++){
            for(let j = 0; j < a.cols; j++){
                result.data[i][j] = a.data[i][j] - b.data[i][j];
            }
        }
        return result;
    }
    
    static multiply(m1, m2){
        if(m1.cols === m2.rows){
            var dotProduct = new Matrix(m1.rows, m2.cols);
            for(let i = 0; i < m1.rows; i++){
                for(let j = 0; j < m2.cols; j++){
                    var sum = 0
                    for(let k = 0; k < m2.rows; k++){
                        sum += m1.data[i][k] * m2.data[k][j];
                    }
                    dotProduct.data[i][j] = sum;
                }
                
            }
            return dotProduct;
        } else {
            throw Error("Columns should match rows of the matrix.")
        }
    }
    
    getMatrix(){
        return this.data;
    }
    
    printTable(){
        console.table(this.getMatrix());
    }
    map(fn){
        let result = new Matrix(this.rows, this.cols);
        for(let i = 0; i < this.rows; i++){
            for(let j = 0; j < this.cols; j++){
                result.data[i][j] = fn(this.data[i][j]);
            }
        }
        return result;
    }
    apply(fn){
        for(let i = 0; i < this.rows; i++){
            for(let j = 0; j < this.cols; j++){
                this.data[i][j] = fn(this.data[i][j]);
            }
        }
    }
    static transpose(matrix){
        let result = new Matrix(matrix.cols, matrix.rows);
        
        for(let i = 0; i < matrix.rows; i++){
            for(let j = 0; j < matrix.cols; j++){
                result.data[j][i] = matrix.data[i][j];
            }
        }
        return result;
    }
    
    add(n){
        if(typeof n !== 'number' && n.getMatrix){
            n = n.getMatrix();
        }
        for(let i = 0; i < this.rows; i++){
            for(let j = 0; j < this.cols; j++){
                if(typeof n === 'number'){
                    this.data[i][j] += n;
                } else {
                    this.data[i][j] += n[i][j];
                }
            }
        }
    }
    
    multiply(n){
        if(typeof n !== 'number' && n.getMatrix){
            n = n.getMatrix();
        }
        for(let i = 0; i < this.rows; i++){
            for(let j = 0; j < this.cols; j++){
                if(typeof n === 'number'){
                    this.data[i][j] *= n;
                } else {
                    this.data[i][j] *= n[i][j];
                }
            }
        }
    }
    
    randomize(n){
        for(let i = 0; i < this.rows; i++){
            for(let j = 0; j < this.cols; j++){
                this.data[i][j] = Math.random() * 2 - 1;
            }
        }
    }
}
export default Matrix;